Added drivers composite – additional drivers feature

pkonopelko opened the pull request:
Task information: https://trello.com/c/ZUvQa6ey/13-drivers-composite

We are aware of issue related to doubled "logger" in composite, but didn't want to postpone the implementation further. We hope that it is a minor issue.

MichalSmiechowicz committed:
Adding gui elements
Co-authored-by: Piotr Robert Konopelko <206060@edu.p.lodz.pl>

MichalSmiechowicz committed:
Finished implementation of composite driver

Co-authored-by: Piotr Robert Konopelko <206060@edu.p.lodz.pl>

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/DriverManager.java:
```diff
@@ -10,18 +11,27 @@
 
+	public synchronized void addDriver(Job2dDriver driver){
+		composite.addDriver(driver);
+		System.out.println(composite);
```
Please, do not print to sysout. If you really need it use logger.

pkonopelko replied:
It was for debugging purposes only, removed in 2a48b9a

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/DriverManager.java:
```diff
@@ -10,18 +11,27 @@
 
 	private Job2dDriver currentDriver = new LoggerDriver();
-
+	private CompositeDriver composite = new CompositeDriver();
```
IMO: CompositeDriver should substitute currentDriver and help driver management but your approach is correct.

pkonopelko replied:
We are not sure how to achieve it. We modified `DriverManager` a little bit in 18cd2d3 to return the composite as a current driver (which is now built earlier, not during the moment of returning it).

We had some ideas to check whether `driversToBeExecuted` contains (or not) a driver while adding or removing it. It partially worked – the logger wasn't added twice, but then we had an issue with checkbox in the additional drivers list that stayed checked in certain circumstances (when logger was disabled – i.e. changed to a different driver – from the main Drivers list).

One idea is to somehow update the status of the checkbox if a driver is "additional", but we currently decided not to dig deeper into it tonight since it's already very late.

Would you share any suggestions with us on that, please? (However can't promise we will take up the gauntlet since we still many many other tasks to complete :))

kowallus replied:
It was only my opinion about the possibility to move more responsibilities to the composite, but your solution is valid.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/SelectAdditionalDriverMenuOptionListener.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        System.out.println(e.paramString());
```
remove

pkonopelko replied:
Done in 2a48b9a

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/SelectAdditionalDriverMenuOptionListener.java:
```diff
@@ -0,0 +1,32 @@
+        }
+        state = !state;
+
+        DriverFeature.updateDriverInfo();
```
now you should integrate your solution with: https://github.com/iis-powp-2020z/powp_jobs2d_projects/pull/2
addDriver and removeDriver implementations should trigger Observer notification automatically.

pkonopelko replied:
Upstream merged in 1d214e7, integration pending

pkonopelko replied:
Integrated with Observer in a4d8ca3

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/adapter/CompositeDriver.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public void operateTo(int i, int i1) {
```
fix names of the params (cryptic)

pkonopelko replied:
Fixed in 2a48b9a

kowallus reviewed (changes requested):
Minor issues to fix + integration with Observer for a driver changes events feature

pkonopelko committed:
Merge remote-tracking branch 'upstream/master'

pkonopelko committed:
Removed printlines, fixed cryptic arguments

* Removed unnecessary printlines
* Refactored arguments names to be less cryptic and more user-friendly in CompositeDriver.java

pkonopelko committed:
Integration with Observer

addDriver and removeDriver implementation triggers Observer notification automatically now

Co-Authored-By: MichalSmiechowicz <56608744+MichalSmiechowicz@users.noreply.github.com>

pkonopelko committed:
Small refactor (spaces)

pkonopelko committed:
Creating Composite moved to setCurrentDriver

Co-Authored-By: MichalSmiechowicz <56608744+MichalSmiechowicz@users.noreply.github.com>

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/DriverManager.java:
```diff
@@ -10,33 +10,39 @@
 	 * @param driver Set the driver as current.
 	 */
 	public synchronized void setCurrentDriver(Job2dDriver driver) {
+		composite.removeDriver(currentDriver);
+		composite.addDriver(driver);
```
nice. Now there are no unnecessary new CompositeDriver(...) calls.

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -11,10 +11,10 @@
-import edu.kis.powp.jobs2d.events.SelectTestFigure2OptionListener;
-import edu.kis.powp.jobs2d.events.SelectTestFigureOptionListener;
+import edu.kis.powp.jobs2d.events.*;
+// import edu.kis.powp.jobs2d.events.SelectRunCurrentCommandOptionListener;
+// import edu.kis.powp.jobs2d.events.SelectTestFigure2OptionListener;
+// import edu.kis.powp.jobs2d.events.SelectTestFigureOptionListener;
```
In revision repo it is safe to delete code instead of using a comment.

pkonopelko replied:
Ahh, it was done just temporarily while merging changes (another group imported whole `edu.kis.powp.jobs2d.events.*`) – it seems that we forgot to remove it later on. Sorry.

kowallus reviewed (approved):
Thanks
